### PR TITLE
GC hazard fix in customelementregistry.rs

### LINKED ISSUE #34017: GC borrow hazard in CustomElementRegistry::WhenDefined
Mutable borrow: https://github.com/servo/servo/blob/5d4b5514aa339d115d0ac9e1131eb7bef1d99a0b/components/script/dom/customelementregistry.rs#L602
Potential GC operation while the borrow is outstanding: https://github.com/servo/servo/blob/5d4b5514aa339d115d0ac9e1131eb7bef1d99a0b/components/script/dom/customelementregistry.rs#L606

We should be able to work around this by calling `self.when_defined.borrow()` and `self.when_defined.borrow_mut()` instead of using the `map` variable.

### PR DESCRIPTION
**Pull request by chickenleaf**
**file:** `servo/components/script/dom/customelementregistry.rs`

**PR Description**

**issue**: borrowed data stored in 'map' variable could cause issues during garbage collection

**changes**: replaced 'map' variable with direct `borrow()` and `borrow_mut()` calls on `self.when_defined`

**impact**: safer memory handling by not storing borrowed data in a variable

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #34017

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because ___

### CODE REVIEW

**Review by jdm — APPROVED** [review]

**Comment by chickenleaf** [comment]
this seems to be failing the MacOS build?

**Comment by chickenleaf** [comment]
> this seems to be failing the MacOS build?

#34022 should fix that 🫡

File: components/script/dom/customelementregistry.rs [hunk 1]
```diff
@@ -598,15 +598,17 @@ impl CustomElementRegistryMethods for CustomElementRegistry {
...
-        });
+        // Steps 3, 4, 5
+        let promise = self
+            .when_defined
+            .borrow()
+            .get(&name)
+            .cloned()
+            .unwrap_or_else(|| {
+                let promise = Promise::new_in_current_realm(comp, can_gc);
+                self.when_defined.borrow_mut().insert(name, promise.clone());
```
Comment (sagudev) [inline components/script/dom/customelementregistry.rs:609]: already borrowed: BorrowMutError (thread Script(1,1), at components/script/dom/customelementregistry.rs:609)

borrow happens at 604

**Review by jdm — APPROVED** [review]
Verified in play.rust-lang.org that this pattern doesn't panic, unlike the previous change. The rules around when borrows end in Rust are quite subtle sometimes!

**Comment by chickenleaf** [comment]
I've been trying to learn that 😭 I have more questions if I can't figure out from the internet I shall get back to u 😭

**Pull request merged**